File: oxidize-server/src/routes/responses.rs
//! Shared response builders for chat/completions endpoints.

use std::sync::Arc;
use std::sync::atomic::AtomicBool;

use axum::{
    Json,
    http::StatusCode,
    response::{
        IntoResponse, Response,
        sse::{Event, KeepAlive, Sse},
    },
};
use futures_util::{StreamExt, stream};
use serde_json::{Value, json};

use crate::runtime::generate::{GenerationError, GenerationResult};

pub fn chat_completion_response(model: String, result: GenerationResult) -> Response {
    (
        StatusCode::OK,
        Json(json!({
            "id": "chatcmpl-oxidize",
            "object": "chat.completion",
            "created": unix_timestamp(),
            "model": model,
            "choices": [
                {
                    "index": 0,
                    "message": { "role": "assistant", "content": result.text },
                    "finish_reason": "stop"
                }
            ],
            "usage": usage_json(result.prompt_tokens, result.completion_tokens)
        })),
    )
        .into_response()
}

pub fn completion_response(model: String, result: GenerationResult) -> Response {
    (
        StatusCode::OK,
        Json(json!({
            "id": "cmpl-oxidize",
            "object": "text_completion",
            "created": unix_timestamp(),
            "model": model,
            "choices": [
                {
                    "index": 0,
                    "text": result.text,
                    "finish_reason": "stop"
                }
            ],
            "usage": usage_json(result.prompt_tokens, result.completion_tokens)
        })),
    )
        .into_response()
}

pub fn chat_completion_stream_response(model: String, content: String) -> Response {
    let model_for_end = model.clone();
    let stream = stream::iter(vec![
        Ok::<Event, std::convert::Infallible>(
            Event::default().data(
                json!({
                    "id": "chatcmpl-oxidize",
                    "object": "chat.completion.chunk",
                    "created": unix_timestamp(),
                    "model": model,
                    "choices": [
                        {
                            "index": 0,
                            "delta": { "role": "assistant", "content": content },
                            "finish_reason": null
                        }
                    ]
                })
                .to_string(),
            ),
        ),
        Ok(Event::default().data(
            json!({
                "id": "chatcmpl-oxidize",
                "object": "chat.completion.chunk",
                "created": unix_timestamp(),
                "model": model_for_end,
                "choices": [{ "index": 0, "delta": {}, "finish_reason": "stop" }]
            })
            .to_string(),
        )),
        Ok(Event::default().data("[DONE]")),
    ]);
    Sse::new(stream)
        .keep_alive(KeepAlive::default())
        .into_response()
}

pub fn completion_stream_response(model: String, text: String) -> Response {
    let model_for_end = model.clone();
    let stream = stream::iter(vec![
        Ok::<Event, std::convert::Infallible>(
            Event::default().data(
                json!({
                    "id": "cmpl-oxidize",
                    "object": "text_completion",
                    "created": unix_timestamp(),
                    "model": model,
                    "choices": [
                        { "index": 0, "text": text, "finish_reason": null }
                    ]
                })
                .to_string(),
            ),
        ),
        Ok(Event::default().data(
            json!({
                "id": "cmpl-oxidize",
                "object": "text_completion",
                "created": unix_timestamp(),
                "model": model_for_end,
                "choices": [{ "index": 0, "text": "", "finish_reason": "stop" }]
            })
            .to_string(),
        )),
        Ok(Event::default().data("[DONE]")),
    ]);
    Sse::new(stream)
        .keep_alive(KeepAlive::default())
        .into_response()
}

/// Token-by-token SSE stream driven by a tokio mpsc channel.
pub fn chat_completion_stream_response_paged(
    model: String,
    rx: tokio::sync::mpsc::Receiver<Result<String, GenerationError>>,
    cancel: Arc<AtomicBool>,
) -> Response {
    let stream = stream::unfold(
        (model.clone(), false, rx, cancel),
        move |(model, done, mut rx, cancel)| async move {
            if done {
                return None;
            }
            match rx.recv().await {
                Some(Ok(piece)) => {
                    let event = Event::default().data(
                        json!({
                            "id": "chatcmpl-oxidize",
                            "object": "chat.completion.chunk",
                            "created": unix_timestamp(),
                            "model": model,
                            "choices": [
                                {
                                    "index": 0,
                                    "delta": { "content": piece },
                                    "finish_reason": null
                                }
                            ]
                        })
                        .to_string(),
                    );
                    Some((
                        Ok::<Event, std::convert::Infallible>(event),
                        (model, false, rx, cancel),
                    ))
                }
                Some(Err(_error)) => {
                    let event = Event::default().data(
                        json!({
                            "id": "chatcmpl-oxidize",
                            "object": "chat.completion.chunk",
                            "created": unix_timestamp(),
                            "model": model,
                            "choices": [{ "index": 0, "delta": {}, "finish_reason": "stop" }]
                        })
                        .to_string(),
                    );
                    Some((Ok(event), (model, true, rx, cancel)))
                }
                None => {
                    let event = Event::default().data(
                        json!({
                            "id": "chatcmpl-oxidize",
                            "object": "chat.completion.chunk",
                            "created": unix_timestamp(),
                            "model": model,
                            "choices": [{ "index": 0, "delta": {}, "finish_reason": "stop" }]
                        })
                        .to_string(),
                    );
                    Some((Ok(event), (model, true, rx, cancel)))
                }
            }
        },
    );

    let done_stream = stream::iter(vec![Ok(Event::default().data("[DONE]"))]);
    Sse::new(stream.chain(done_stream))
        .keep_alive(KeepAlive::default())
        .into_response()
}

pub fn usage_json(prompt_tokens: usize, completion_tokens: usize) -> Value {
    json!({
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "total_tokens": prompt_tokens.saturating_add(completion_tokens)
    })
}

pub fn generation_error_response(error: GenerationError) -> Response {
    match error {
        GenerationError::KvCacheExhausted => (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({
                "error": { "message": "KV cache exhausted", "type": "insufficient_memory" }
            })),
        )
            .into_response(),
        GenerationError::Other(msg) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": {"message": msg, "type": "generation_error"}})),
        )
            .into_response(),
    }
}

pub fn validate_candidate_count(n: Option<usize>, best_of: Option<usize>) -> Option<Response> {
    let n = n.unwrap_or(1);
    let best_of = best_of.unwrap_or(n);
    if n == 1 && best_of == 1 {
        return None;
    }
    Some(
        (
            StatusCode::BAD_REQUEST,
            Json(json!({
                "error": {
                    "message": "oxidize-server currently supports only n=1 and best_of=1",
                    "type": "unsupported_parameter"
                }
            })),
        )
            .into_response(),
    )
}

pub fn model_not_found(model: &str) -> Response {
    (
        StatusCode::NOT_FOUND,
        Json(json!({
            "error": {
                "message": format!("model '{model}' is not loaded"),
                "type": "model_not_found"
            }
        })),
    )
        .into_response()
}

pub fn unix_timestamp() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_or(0, |duration| duration.as_secs())
}

File: oxidize-golang/internal/api/responses.go
package api

type ErrorResponse struct {
	StatusCode int      `json:"-"`
	Error      APIError `json:"error"`
}

type APIError struct {
	Message string  `json:"message"`
	Type    string  `json:"type"`
	Code    *string `json:"code,omitempty"`
}

type Usage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

func placeholderUsage() Usage {
	return Usage{PromptTokens: 0, CompletionTokens: 0, TotalTokens: 0}
}

type ModelsResponse struct {
	Object string      `json:"object"`
	Data   []ModelData `json:"data"`
}

type ModelData struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	OwnedBy string `json:"owned_by"`
}

type ChatCompletionResponse struct {
	ID      string       `json:"id"`
	Object  string       `json:"object"`
	Created int64        `json:"created"`
	Model   string       `json:"model"`
	Choices []ChatChoice `json:"choices"`
	Usage   Usage        `json:"usage"`
}

type ChatChoice struct {
	Index        int          `json:"index"`
	Message      *ChatMessage `json:"message,omitempty"`
	Delta        *ChatDelta   `json:"delta,omitempty"`
	FinishReason *string      `json:"finish_reason"`
}

type ChatDelta struct {
	Role    string `json:"role,omitempty"`
	Content string `json:"content,omitempty"`
}

type TextCompletionResponse struct {
	ID      string       `json:"id"`
	Object  string       `json:"object"`
	Created int64        `json:"created"`
	Model   string       `json:"model"`
	Choices []TextChoice `json:"choices"`
	Usage   Usage        `json:"usage"`
}

type TextChoice struct {
	Index        int     `json:"index"`
	Text         string  `json:"text"`
	FinishReason *string `json:"finish_reason"`
}

type EmbeddingsResponse struct {
	Object string          `json:"object"`
	Data   []EmbeddingData `json:"data"`
	Model  string          `json:"model"`
	Usage  EmbeddingsUsage `json:"usage"`
}

type EmbeddingData struct {
	Object    string    `json:"object"`
	Embedding []float64 `json:"embedding"`
	Index     int       `json:"index"`
}

type EmbeddingsUsage struct {
	PromptTokens int `json:"prompt_tokens"`
	TotalTokens  int `json:"total_tokens"`
}

func BuildChatCompletion(model string, content string) ChatCompletionResponse {
	stop := "stop"
	return ChatCompletionResponse{
		ID: "chatcmpl-placeholder", Object: "chat.completion", Created: 0, Model: model,
		Choices: []ChatChoice{{Index: 0, Message: &ChatMessage{Role: "assistant", Content: NewMessageContentText(content)}, FinishReason: &stop}},
		Usage:   placeholderUsage(),
	}
}

func BuildChatChunk(model string, content string, finished bool) ChatCompletionResponse {
	var finish *string
	choice := ChatChoice{Index: 0}
	if finished {
		stop := "stop"
		finish = &stop
		choice.Delta = &ChatDelta{}
	} else {
		choice.Delta = &ChatDelta{Role: "assistant", Content: content}
	}
	choice.FinishReason = finish
	return ChatCompletionResponse{ID: "chatcmpl-placeholder", Object: "chat.completion.chunk", Created: 0, Model: model, Choices: []ChatChoice{choice}}
}

func BuildTextCompletion(model string, text string) TextCompletionResponse {
	stop := "stop"
	return TextCompletionResponse{
		ID: "cmpl-placeholder", Object: "text_completion", Created: 0, Model: model,
		Choices: []TextChoice{{Index: 0, Text: text, FinishReason: &stop}},
		Usage:   placeholderUsage(),
	}
}

func BuildTextChunk(model string, text string, finished bool) TextCompletionResponse {
	var finish *string
	out := text
	if finished {
		stop := "stop"
		finish = &stop
		out = ""
	}
	return TextCompletionResponse{ID: "cmpl-placeholder", Object: "text_completion", Created: 0, Model: model, Choices: []TextChoice{{Index: 0, Text: out, FinishReason: finish}}, Usage: placeholderUsage()}
}

func BuildModelsResponse(modelIDs ...string) ModelsResponse {
	if len(modelIDs) == 0 {
		modelIDs = []string{"oxidize-default"}
	}
	data := make([]ModelData, 0, len(modelIDs))
	for _, modelID := range modelIDs {
		data = append(data, ModelData{ID: modelID, Object: "model", OwnedBy: "oxidize"})
	}
	return ModelsResponse{Object: "list", Data: data}
}

func BuildEmbeddingsResponse(model string) EmbeddingsResponse {
	embedding := make([]float64, 8)
	return EmbeddingsResponse{
		Object: "list",
		Data:   []EmbeddingData{{Object: "embedding", Embedding: embedding, Index: 0}},
		Model:  model,
		Usage:  EmbeddingsUsage{},
	}
}

func ValidateCandidateCount(n *int, bestOf *int) *ErrorResponse {
	actualN := 1
	actualBestOf := 1
	if n != nil {
		actualN = *n
		actualBestOf = actualN
	}
	if bestOf != nil {
		actualBestOf = *bestOf
	}
	if actualN == 1 && actualBestOf == 1 {
		return nil
	}
	return &ErrorResponse{StatusCode: 400, Error: APIError{Message: "oxidize-server currently supports only n=1 and best_of=1", Type: "unsupported_parameter"}}
}

func ModelNotFound(model string) ErrorResponse {
	return ErrorResponse{StatusCode: 404, Error: APIError{Message: "model '" + model + "' is not loaded", Type: "model_not_found"}}
}

func InvalidAPIKey() ErrorResponse {
	code := "invalid_api_key"
	return ErrorResponse{StatusCode: 401, Error: APIError{Message: "Incorrect API key provided", Type: "invalid_request_error", Code: &code}}
}

func MalformedJSON() ErrorResponse {
	return ErrorResponse{StatusCode: 400, Error: APIError{Message: "malformed json", Type: "invalid_request_error"}}
}
